fleetspeak/src/server/components/components.go

```diff
@@ -139,14 +139,11 @@ func MakeComponents(cfg cpb.Config) (*server.Components, error) {
 	scfg := cfg.StatsConfig
 	var statsCollector stats.Collector
 	if scfg != nil {
-		portToExportStats := scfg.Port
-		if portToExportStats == "" {
-			portToExportStats = "2112" // default port
-		}
-		if scfg.Collector == "prometheus" {
+		addressToExportStats := scfg.Address
+		if addressToExportStats != "" {
 			statsCollector = server.PrometheusStatsCollector{}
 			http.Handle("/metrics", promhttp.Handler())
-			go http.ListenAndServe(":" + portToExportStats, nil)
+			go http.ListenAndServe(addressToExportStats, nil)
```

Comment:
I just realize that we need to figure out if this use of the http global ServerMux interacts with the http server that is run over in the notification code: https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/components/notifications/http.go

Skimming to docs, I think both handlers must end up on the same DefaultServeMux, and then I'm really not sure what running Serve twice at the same time does.

The two ways to clean this up are 1) create actual http.Server objects in both cases, so that they end up with separate multiplexers and start/stop independently, or 2) organize a common http server for internal coordination and metrics (possibly based on the default mux). configure its bind address in one place and attach both handlers to it.

Reply:
> I just realize that we need to figure out if this use of the http global ServerMux interacts with the http server that is run over in the notification code: https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/components/notifications/http.go
> 
> Skimming to docs, I think both handlers must end up on the same DefaultServeMux, and then I'm really not sure what running Serve twice at the same time does.
> 
> The two ways to clean this up are 1) create actual http.Server objects in both cases, so that they end up with separate multiplexers and start/stop independently, or 2) organize a common http server for internal coordination and metrics (possibly based on the default mux). configure its bind address in one place and attach both handlers to it.

Me and @mbushkov have taken a look into [this article](https://echorand.me/posts/golang-dissecting-listen-and-serve/#using-your-own-handler-with-listenandserve) which suggested to create a separate mux other than DefaultServeMux and attach the Prometheus HTTP handler to it.
Please have a look at 4c85c00.

Reply:
Yeah, that fixes the problem locally, and is good enough to get started with. Long term we should think about whether we want one or two ports for these internal communications purposes. (there is also a more involved example alread in the code at https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/https/https.go#L95)